# Immunology Data and Repositories

a.k.a. Where can I get imunological data and where can I put my data

Outline
- Data Standards
  - What do I mean by standards?
  - Why have standards
  - How do data standards benefit you?
Reproducible science
Data Repositories
What are data repositories?
Why have them
How do they benefit you?
How can you contribute data?
How can you find them?
Resource reposotories






























Table 1: Analysis and visualization tools for systems immunology
Data types or tools	Program name	Weblink
TCR repertoire	MiTCR36	http://mitcr.milaboratory.com/
 	Decombinator101	https://github.com/uclinfectionimmunity/Decombinator
 	iSSAKE102	ftp://ftp.bcgsc.ca/supplementary/iSSAKE/
 	IMGT or HIGHV-Quest37	http://www.imgt.org/HighV-QUEST/
 	 	 
Antibody repertoire	IMGT or HIGHV-Quest37	http://www.imgt.org/HighV-QUEST/
 	IgTree103	http://immsilico2.lnx.biu.ac.il/Software.html
 	VDJFasta104	http://sourceforge.net/projects/vdjfasta/
 	 	 
RNA-seq105	VarScan106	http://varscan.sourceforge.net/
 	GATK107	http://www.broadinstitute.org/gatk/
 	SAMtools108	http://samtools.sourceforge.net/
 	ERANGE40	http://woldlab.caltech.edu/rnaseq
 	Scripture109	http://www.broadinstitute.org/software/scripture/
 	Cufflinks45	http://cufflinks.cbcb.umd.edu/
 	CuffDiff45	http://cufflinks.cbcb.umd.edu/
 	EdgeR46	http://www.bioconductor.org/packages/2.12/bioc/html/edgeR.html
 	DESeq49	http://www.bioconductor.org/packages/release/bioc/html/DESeq.html
 	Myrna110	http://bowtie-bio.sourceforge.net/myrna/index.shtml
 	PoissonSeq48	http://cran.r-project.org/web/packages/PoissonSeq/index.html
 	 	 
ChIP-seq105, 111	ERANGE40	http://woldlab.caltech.edu/rnaseq
 	CisGenome112	http://www.biostat.jhsph.edu/~hji/cisgenome/
 	MACS113	http://liulab.dfci.harvard.edu/MACS/
 	PeakSeq114	http://info.gersteinlab.org/PeakSeq/
 	SPP115	http://compbio.med.harvard.edu/Supplements/ChIP-seq/
 	 	 
eQTL	MatrixEQTL58	http://www.bios.unc.edu/research/genomic_software/Matrix_eQTL/
 	PLINK116	http://pngu.mgh.harvard.edu/~purcell/plink/
 	R/qtl117	http://www.rqtl.org/
 	snpMatrix118	http://www.bioconductor.org/packages/2.3/bioc/html/snpMatrix.html
 	 	 
Networks	WGCNA82	http://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/
 	coXpress119	http://coxpress.sourceforge.net/
 	Inferelator120	http://bonneaulab.bio.nyu.edu/networks.html
 	ARACNE121	http://wiki.c2b2.columbia.edu/workbench/index.php/ARACNe/
 	RimbaNET87	http://icahn.mssm.edu/departments-and-institutes/genomics/about/software/rimbanet/
 	 	 
Cytometry63	FlowMeans122	http://www.bioconductor.org/packages/release/bioc/html/flowMeans.html
 	FLAME123	http://www.broadinstitute.org/cancer/software/genepattern/modules/FLAME/
 	FLOCK124	https://immport.niaid.nih.gov/immportWeb/analysis/flow/displayOverview.do?method=displayFlockOverview
 	SamSPECTRAL125	http://www.bioconductor.org/packages/devel/bioc/html/SamSPECTRAL.html
 	SPADE64	http://www.bioconductor.org/packages/release/bioc/html/spade.html
 	 	http://cytospade.org/
 	viSNE66	http://www.c2b2.columbia.edu/danapeerlab/html/cyt.html
 	 	 
Visualization	Cytoscape93	http://www.cytoscape.org/
 	Gephi	https://gephi.org/
 	Circos126	http://circos.ca/
 	 	 
Databases or online tools	ImMunoGeneTics127	http://www.imgt.org/
 	ImmPort	https://immport.niaid.nih.gov/
 	MsigDB53	http://www.broadinstitute.org/gsea/msigdb/index.jsp
 	GenePattern54	http://www.broadinstitute.org/cancer/software/genepattern/
 	DAVID55, 56	http://david.abcc.ncifcrf.gov/
 	Stanford Data Miner128	 
 	Structure129	http://pritchardlab.stanford.edu/structure.html
 	PCAdmix	https://sites.google.com/site/pcadmix/home
 	myPEG130	http://www.megasoftware.net/myPEG/myPEG.php
 	GWASdb131	http://jjwanglab.org/gwasdb/
 	HaploReg132	http://www.broadinstitute.org/mammals/haploreg/haploreg.php
 	RegulomeDB133	http://regulomedb.org/
 	GRAIL134	http://www.broadinstitute.org/mpg/grail/
 	DGIdb135	http://dgidb.genome.wustl.edu/
 	GSEA50	http://www.broadinstitute.org/gsea/index.jsp
 	ProfileChaser136	http://profilechaser.stanford.edu/
 	LINCS browser137, 138	http://lincs.hms.harvard.edu/explore/pathway/
 	 	http://lincs.hms.harvard.edu/explore_/canvasbrowser/
